pkg/binary/envoy/debug/lsof.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2019 Tetrate
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package debug
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"syscall"
+
+	"github.com/shirou/gopsutil/process"
+	"github.com/tetratelabs/getenvoy/pkg/binary"
+	"github.com/tetratelabs/getenvoy/pkg/binary/envoy"
+	"github.com/tetratelabs/log"
+)
+
+// Lsof defines the structure of statistics about a single opened file
+type Lsof struct {
+	// string enclosed in `` are known as struct tags
+	// this particular tag is used by json.Marshal() to encode Command field in a specific manner
+	// process dependent information
+	Command string `json:"command"`
+	Pid     string `json:"pid"`
+	User    string `json:"user"`
+	Fd      string `json:"fd"`
+	// non process dependent information
+	Type   string `json:"type"`   // type of node associated with the file / FIFO for FIFO special file
+	Device string `json:"device"` // device numbers associated with the file, separated by commas
+	Size   string `json:"size"`
+	Node   string `json:"node"` // inode number of a local file/ Internet protocol type
+	Name   string `json:"name"` // name of the mount point and file system on which the file resides
+}
+
+// EnableOpenFilesDataCollection is a preset option that registers collection of statistics of files opened by envoy instance(s)
+func EnableOpenFilesDataCollection(r *envoy.Runtime) {
+	if err := os.Mkdir(filepath.Join(r.DebugStore(), "lsof"), os.ModePerm); err != nil {
+		log.Errorf("error in creating a directory to write open file data of envoy to: %v", err)
+	}
+	r.RegisterPreTermination(retrieveOpenFilesData)
+}
+
+// retrieveOpenFilesData writes statistics of open files associated with envoy instance(s) to a json file
+// if succeeded, return nil, else return an error instance
+func retrieveOpenFilesData(r binary.Runner) error { //nolint:gocyclo
+	// get a list of processes
+	processes, err := process.Processes()
+	if err != nil {
+		return fmt.Errorf("error in getting process pids")
+	}
+
+	// filter Process instances of envoy
+	isEnvoy := func(p *process.Process) bool {
+		name, err := p.Name() //nolint:govet
```

Comment:
whats the complaint here?

Reply:
@liamawhite The error is ` shadow: declaration of "err" shadows declaration at line 59`. I think the error is returned because I reused the identifier `err` in another function scope. Would it be better if I just change the variable name instead of ignoring the lint?

Reply:
yes, avoid shadowing